src/librustc/lint.rs

```diff
@@ -214,92 +234,122 @@ pub fn struct_lint_level<'a>(
+    let future_incompatible = lint.future_incompatible;
+    if let Some(future_incompatible) = future_incompatible {
+        if lint_id == LintId::of(crate::lint::builtin::UNSTABLE_NAME_COLLISIONS) {
+            err.warn(
+                "once this method is added to the standard library, \
+                      the ambiguity may cause an error or change in behavior!",
+            );
         } else if lint_id == LintId::of(builtin::MUTABLE_BORROW_RESERVATION_CONFLICT) {
-            "this borrowing pattern was not meant to be accepted, \
-             and may become a hard error in the future"
-                .to_owned()
-        } else if let Some(edition) = future_incompatible.edition {
-            format!("{} in the {} edition!", STANDARD_MESSAGE, edition)
+            err.warn(
+                "this borrowing pattern was not meant to be accepted, \
+                      and may become a hard error in the future",
+            );
         } else {
-            format!("{} in a future release!", STANDARD_MESSAGE)
-        };
-        let citation = format!("for more information, see {}", future_incompatible.reference);
-        err.warn(&explanation);
-        err.note(&citation);
+            let hard_err_msg = if let Some(edition) = future_incompatible.edition {
+                format!("it will become a hard error in the {} edition!", edition)
+            } else {
+                format!("it will become a hard error in a future release!")
+            };
+
+            let previously_msg = if let Some(n) = name {
+                format!(
+                    "`{}` was previously accepted by the compiler but is being phased out; {}",
+                    n, hard_err_msg
+                )
+            } else {
+                format!(
+                    "this was previously accepted by the compiler but is being phased out; {}",
+                    hard_err_msg
+                )
+            };
+            err.warn(&previously_msg);
+        }
+
+        err.note(&format!("for more information, see {}", future_incompatible.reference));
     }
 
-    return err;
+    // If this code originates in a foreign macro, aka something that this crate
+    // did not itself author, then it's likely that there's nothing this crate
+    // can do about it. We probably want to skip the lint entirely.
+    if err.span.primary_spans().iter().any(|s| in_external_macro(sess, *s)) {
+        // Any suggestions made here are likely to be incorrect, so anything we
+        // emit shouldn't be automatically fixed by rustfix.
+        err.allow_suggestions(false);
+
+        // If this is a future incompatible lint it'll become a hard error, so
+        // we have to emit *something*. Also allow lints to whitelist themselves
+        // on a case-by-case basis for emission in a foreign macro.
+        if future_incompatible.is_none() && !lint.report_in_external_macro {
+            err.cancel()
+        }
+    }
```

Comment:
Bad rebase?

Reply:
I'm not sure what you mean.

Reply:
Looking at the diff, it seems like this code is now duplicated from earlier in `struct_span_level` (take a look at "view file"), so I'm wondering if you inadvertently brought old code in with the forward port.